File: 0020 Valid Parentheses/solution.cpp
#include <bits/stdc++.h>

using std::string; using std::stack;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{')
                stk.push(s[i]);
            else if (s[i] == ')') {
                if (stk.empty() || stk.top() != '(')
                    return false;
                stk.pop();
            }
            else if (s[i] == ']') {
                if (stk.empty() || stk.top() != '[')
                    return false;
                stk.pop();
            }
            else {
                if (stk.empty() || stk.top() != '{')
                    return false;
                stk.pop();
            }
        }
        return stk.empty();
    }
};

int main(){
    Solution s;
    string s1 = "()[]{}";
    assert(s.isValid(s1) == true);
    string s2 = "(]";
    assert(s.isValid(s2) == false);
    return 0;
}

File: 0020 Valid Parentheses/solution.py
#! python3

class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if not s:    return True
        stack = []
        for c in s:
            if c in ['(', '[', '{']:
                stack.append(c)
            else:
                if stack:
                    t = stack.pop()
                else:
                    return False
                
                if (t, c) in [('(', ')'), ('[', ']'), ('{', '}')]:
                    continue
                else:
                    return False
        
        return len(stack)==0

if __name__ == "__main__":
    print(Solution().isValid("()[]{}"))
    print(Solution().isValid("(])"))